go/vt/mysqlctl/cmd.go

```diff
@@ -30,6 +30,35 @@ import (
 // installation that hasn't been set up yet.
 func CreateMysqldAndMycnf(tabletUID uint32, mysqlSocket string, mysqlPort int32) (*Mysqld, *Mycnf, error) {
 	mycnf := NewMycnf(tabletUID, mysqlPort)
+
+	// Choose a random MySQL server-id, since this is a fresh data dir.
+	// We don't want to use the tablet UID as the MySQL server-id,
+	// because reusing server-ids is not safe.
+	//
+	// For example, if a tablet comes back with an empty data dir, it will restore
+	// from backup and then connect to the primary. But if this tablet has the same
+	// server-id as before, and if this tablet was recently a primary, then it can
+	// lose data by skipping binlog events due to replicate-same-server-id=FALSE,
+	// which is the default setting.
+	if err := mycnf.RandomizeMysqlServerID(); err != nil {
+		return nil, nil, fmt.Errorf("couldn't generate random MySQL server_id: %v", err)
+	}
+	if mysqlSocket != "" {
+		mycnf.SocketFile = mysqlSocket
+	}
+
+	dbconfigs.GlobalDBConfigs.InitWithSocket(mycnf.SocketFile)
+	return NewMysqld(&dbconfigs.GlobalDBConfigs), mycnf, nil
+}
+
+// CreateMysqldAndMycnf returns a Mysqld and a Mycnf object to use for working with a MySQL
+// installation that hasn't been set up yet.
+func CreateMysqldAndMycnfFromSources(tabletUID uint32, mysqlSocket string, mysqlPort int32) (*Mysqld, *Mycnf, error) {
+	mycnf, err := NewMycnfFromSources(tabletUID, mysqlPort)
+	if err != nil {
+		return nil, nil, err
+	}
+
```

Comment:
How are the changes in this file related to the PR? I don't see it mentioned anywhere as an issue that the PR is meant to address. Is this needed because of other work added in this PR? Or perhaps it's related to the new vitess pflag work?

Reply:
All of the CNF stuff in this PR is a side-effect of trying to establish a deterministic log file location, in order to answer the E2E test question "did MySQL disable and then redo the redo log during its run".

Reply:
To say a bit more about the CNF challenges:
 * Throughout the code base there are different ways to create a MySQL CNF:
   * Create an empty MyCNF, don't use `--flags` and on-disk path.
   * Using only mysqlctl `--flags`.
   * Using a combination of `--mycnf-path` and the contents of that path.
   * Using a tablet-default MyCNF path and the contents of that path.
 * Prior to the PR `vtbackup` [uses the "create empty MyCNF" variant](https://github.com/vitessio/vitess/blob/55695d1/go/cmd/vtbackup/vtbackup.go#L253).
 * With this PR, I wanted a way to be able to tell `vtbackup` where to store the MySQL error log. In order to do that, I reworked some of the mysqlctl CNF-loading code.
   * You might be asking yourself, "if you can tell vtbackup to store the error log at `--mycnf_error_log_path`, why do you even need `--tablet-path`?" For the life of me I cannot remember if I just didn't think of this until now, or if I ever had a good reason for it, but at the moment, it seems to me like I could just rely purely on setting `--mycnf_error_log_path` from the E2E test, and do without `--tablet-path`.